### PR TITLE
feat: unify js dependency

### PR DESCRIPTION
**Pull request by SyMind**
## Summary

feat: unify js dependency

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by SyMind** [comment]
!bench

**Comment by SyMind** [comment]
!eco-ci

File: crates/rspack_binding_values/src/dependency.rs [hunk 1]
```diff
@@ -1,108 +1,193 @@
...
+      self.compilation = Some(compilation);
+    }
   }
 
-  #[napi(getter)]
-  pub fn category(&self) -> &str {
-    self.0.category().as_str()
-  }
+  fn as_ref(&mut self) -> napi::Result<&dyn Dependency> {
+    let dependency = unsafe { self.dependency.as_ref() };
```
Comment (JSerFeng) [inline crates/rspack_binding_values/src/dependency.rs:26]: safety comments ?

Comment (SyMind) [reply crates/rspack_binding_values/src/dependency.rs:26]: The JS API is now in a fixed pattern and will use ids to verify pointers, if invalid will return Err. So I think safe comments are unnecessary.

**Comment by JSerFeng** [comment]
This pr depends on accessing a field within a raw pointer, for example

```
// ptr is raw pointer and might point to whatever possible
let foo = ptr.as_ref();

foo.id == last_id
```

But I wonder if access id here will panic ? And this is a solution that relies on UB ? Can we accept this @h-a-n-a

**Comment by h-a-n-a** [comment]
> But I wonder if access id here will panic ? And this is a solution that relies on UB ? Can we accept this @h-a-n-a

No, this will not panic and it's more likely a undefined behavior if the memory you're accessing to is freed and it will likely to cause segmentation faults. All we need to do is to prevent user from accessing that value after the scope is dropped.

**Review by JSerFeng — APPROVED** [review]

**Pull request merged**